File: src/main/java/com/busted_moments/client/models/war/War.java
package com.busted_moments.client.models.war;

import com.busted_moments.core.api.requests.mapstate.Territory;
import com.busted_moments.core.time.Duration;
import com.busted_moments.core.time.ChronoUnit;

import java.util.*;

public final class War {
   private final Territory territory;
   private final Date enteredAt;
   Date startedAt;
   Tower tower;

   private Date endedAt = null;

   public War(Territory territory, Date enteredAt) {
      this.territory = territory;
      this.enteredAt = enteredAt;
   }

   public Territory getTerritory() {
      return territory;
   }

   public Date getEnteredAt() {
      return enteredAt;
   }

   public Date getStartedAt() {
      return startedAt;
   }

   public Date getEndedAt() {
      return endedAt;
   }

   public Tower getTower() {
      return tower;
   }

   public double getDPS(Duration duration) {
      if (tower == null || tower.size() < 2) return 0;

      if (duration.isForever()) return (tower.getInitialStats().ehp() - tower.getStats().ehp()) / Duration.since(getStartedAt()).toSeconds();

      final Tower.Stats[] points = {null, null};

      tower.stream()
              .filter(update -> Duration.since(update.date()).lessThanOrEqual(duration))
              .forEach(update -> {
                 if (points[0] == null) points[0] = update.before();
                 points[1] = update.after();
              });

      if (points[0] == null || points[1] == null) return 0;
      return (points[0].ehp() - points[1].ehp()) / duration.toSeconds();
   }

   public double getDPS(double length, ChronoUnit unit) {
      return getDPS(Duration.of(length, unit));
   }

   public Duration getDuration() {
      if (startedAt == null) return Duration.of(0, ChronoUnit.SECONDS);

      return Duration.since(startedAt);
   }

   public boolean hasStarted() {
      return startedAt != null;
   }

   public boolean hasEnded() {
      return endedAt != null;
   }

   void end() {
      if (!hasEnded()) endedAt = new Date();
   }

   @Override
   public String toString() {
      return "War{" +
              "territory=" + territory +
              ", enteredAt=" + enteredAt +
              ", startedAt=" + startedAt +
              ", tower=" + tower +
              '}';
   }
}

File: src/main/kotlin/com/busted_moments/client/models/territories/war/War.kt
package com.busted_moments.client.models.territories.war

import com.busted_moments.buster.api.GuildType
import com.busted_moments.buster.api.Territory
import com.busted_moments.client.buster.GuildList
import com.busted_moments.client.framework.events.post
import com.busted_moments.client.models.territories.war.events.WarEvent
import net.essentuan.esl.json.Json
import net.essentuan.esl.model.annotations.Ignored
import net.essentuan.esl.time.duration.Duration
import net.essentuan.esl.time.extensions.timeSince
import java.util.Date
import java.util.UUID

class War(
    val territory: Territory,
    val enteredAt: Date
) {
    lateinit var tower: Tower
        internal set
    lateinit var startedAt: Date
        internal set
    lateinit var endedAt: Date
        internal set

    val hasEnded: Boolean
        get() = ::endedAt.isInitialized

    val hasStarted: Boolean
        get() = ::startedAt.isInitialized

    val active: Boolean
        get() = hasStarted && !hasEnded

    fun update(
        stats: Tower.Stats,
        at: Date = Date()
    ) {
        if (!hasStarted) {
            startedAt = Date()

            val update = Tower.Update(
                at,
                stats,
                stats
            )

            tower = Tower(mutableListOf(update))
            WarEvent.Start(this).post()
            WarEvent.TowerUpdate(this, update).post()
        } else if (active)
            tower.update(this, stats, at)
    }

    fun dps(duration: Duration): Double = when {
        !::startedAt.isInitialized || !::tower.isInitialized || tower.size < 2 -> 0.0
        duration.isForever -> (tower.initial.ehp - tower.stats.ehp) / startedAt.timeSince().toSeconds()
        else -> {
            var first: Tower.Stats? = null
            var last: Tower.Stats? = null

            for (update in tower) {
                if (update.at.timeSince() > duration)
                    continue

                if (first == null)
                    first = update.before

                last = update.after
            }

            if (first == null || last == null)
                0.0
            else
                (first.ehp - last.ehp) / duration.toSeconds()
        }
    }

    data class Results(
        val started: Date,
        val ended: Date,
        val territory: String,
        private val uuid: UUID,
        val initial: Tower.Stats,
        val final: Tower.Stats
    ) : Json.Model {
        @Ignored
        private var _owner: GuildType? = null

        val duration: Duration
            get() = Duration(started, ended)

        val dps: Double
            get() =
                ((initial.ehp - final.ehp) / duration.toSeconds().let { if (it == 0.0) 1.0 else it })

        val owner: GuildType
            get() {
                val owner = _owner

                if (owner != null)
                    return owner
                else {
                    val guild = GuildList[uuid]

                    if (guild != null) {
                        _owner = guild

                        return guild
                    }
                }

                return object : GuildType {
                    override val name: String
                        get() = "Unknown"
                    override val tag: String
                        get() = "----"
                    override val uuid: UUID
                        get() = uuid

                }
            }
    }
}